### PR TITLE
OTAServerDelegate class for vendor logic

### LINKED ISSUE #2862: Implement OTA Software Update cluster (Server side)
#### Problem
The CHIP spec defines an OTA software update cluster but the SDK currently does not implement said cluster's server side logic

 #### Proposed Solution
Implement the CHIP OTA software update cluster's server side logic

### PR DESCRIPTION
**Pull request by holbrookt**
#### Problem
Need a way to separate common code from application/vendor specific code for OTA Server Cluster, which doesn't require the developer to fork the repo.

#### Change overview
* define OTAServerDelegate class for implementing vendor OTA Server logic
* add some common logic for OTA Server Cluster commands
* add OTA Server Cluster to all-clusters-app and chip-tool
* fix AnnounceOTAServer command direction (OTA Client client node -> OTA Client server node)
* Fixes #2862 - Implement OTA Software Update cluster (Server side)

#### Testing
Ran `chip-tool` and `all-clusters-app` on Linux machine, confirmed that I can send `query-image` `apply-update-request` and `notify-update-applied`. The response to `apply-update-request` and `notify-update-applied` is `EMBER_ZCL_STATUS_UNSUP_COMMAND` as expected, because there is no OTAServerDelegate implemented yet. The response to `query-image` is a TLV parse failure because `CHAR_STRING` is not supported yet

### CODE REVIEW

**Comment by holbrookt** [comment]
Will add generated files when review of edited files is complete

File: examples/all-clusters-app/all-clusters-common/BUILD.gn
Comment (holbrookt) [inline examples/all-clusters-app/all-clusters-common/BUILD.gn:25]: Is this an acceptable way to conditionally compile the example OTA Server code?

File: src/app/clusters/ota-server/ota-server.cpp
Comment (holbrookt) [inline src/app/clusters/ota-server/ota-server.cpp:61]: Is `EMBER_ZCL_STATUS_INVALID_FIELD` the correct error to return for when a parameter is unexpected length?

Comment (woody-apple) [reply src/app/clusters/ota-server/ota-server.cpp:61]: @vivien-apple ?

File: src/app/clusters/ota-client/ota-client.cpp
Comment (tcarmelveilleux) [inline src/app/clusters/ota-client/ota-client.cpp:47]: Wrong log

File: src/app/clusters/ota-server/ota-server-delegate.h
Comment (tcarmelveilleux) [inline src/app/clusters/ota-server/ota-server-delegate.h:33]: Why are all these methods static? Making them static implies some wider singleton definition than just having a single instance

File: src/app/clusters/ota-server/ota-server.cpp
Comment (tcarmelveilleux) [inline src/app/clusters/ota-server/ota-server.cpp:138]: The API should allow different implementations to be used, which may embed different state-keeping. Suggest making the OTAServerDelegate a virtual interface.

Comment (holbrookt) [reply src/app/clusters/ota-server/ota-server.cpp]: Agreed. I made the Delegate methods virtual and added a function for setting the Delegate for the ember callbacks to use

**Comment by woody-apple** [comment]
@holbrookt Can you take a peek at the failing ZAP task?

**Comment by holbrookt** [comment]
> @holbrookt Can you take a peek at the failing ZAP task?

It's failing because I pushed edited ZAP files and didn't include the generated ones. I was going to wait to include the generated ones until the end of the PR approval process to avoid merge conflicts. But maybe people are waiting to review until CI is green?

**Review by woody-apple — CHANGES_REQUESTED** [review]
Per the updated template, can you update the PR here?

```
#### Problem
What is being fixed?  Examples:
* Fix crash on startup
* Fixes #12345 12345 Frobnozzle is leaky (exactly like that, so GitHub will auto-close the issue).

#### Change overview
What's in this PR

#### Testing
How was this tested? (at least one bullet point required)
    • If unit tests were added, how do they cover this issue?
    • If unit tests existed, how were they fixed/modified to prevent this in future?
    • If integration tests were added, how do they verify this change?
    • If manually tested, what platforms controller and device platforms were manually tested, and how?
    • If no testing is required, why not?
```

**Comment by holbrookt** [comment]
PR description has been fixed to adhere to the new template.

Apologies for the messy commit history, I've been struggling with weird zap configure tool behavior, and missing zap files from an existing PR (?).

File: src/app/clusters/ota-server/ota-server.cpp
Comment (tcarmelveilleux) [reply src/app/clusters/ota-server/ota-server.cpp:75]: CHIP spec has a new status code:
```
| INVALID_ARGUMENT        | 0xc6  | Argument(s) specified in a command are invalid for that particular command invocation
```

It is possible it is not defined yet. You could check.

File: src/app/clusters/ota-server/ota-server.cpp
Comment (tcarmelveilleux) [inline src/app/clusters/ota-server/ota-server.cpp:79]:
```suggestion
    if (SendStatusIfDelegateNull())
    {
        return true;
    }
```

File: src/app/clusters/ota-server/ota-server.cpp
Comment (tcarmelveilleux) [inline src/app/clusters/ota-server/ota-server.cpp:107]: This should probably be INVALID_ARGUMENT

File: src/app/clusters/ota-server/ota-server.cpp
Comment (tcarmelveilleux) [inline src/app/clusters/ota-server/ota-server.cpp:149]: Why is this check there: `locationLen != kLocationParamLength` ? Are variable length strings not yet supported in commands?

File: src/app/clusters/ota-server/ota-server.cpp
Comment (tcarmelveilleux) [inline src/app/clusters/ota-server/ota-server.cpp:157]:
```suggestion
    if (SendStatusIfDelegateNull())
    {
        return true;
    }
```

File: src/app/clusters/ota-server/ota-server.cpp
Comment (tcarmelveilleux) [inline src/app/clusters/ota-server/ota-server.cpp:170]: Any reason why this is global? Is it because the framework doesn't allow cleanly associating particular arguments/delegates to given cluster instances?

File: src/app/clusters/ota-server/ota-server.cpp
Comment (holbrookt) [reply src/app/clusters/ota-server/ota-server.cpp:79]: good catch, thanks

File: src/app/clusters/ota-server/ota-server.cpp
Comment (holbrookt) [reply src/app/clusters/ota-server/ota-server.cpp:107]: There is no EMBER_ZCL_STATUS_INVALID_ARGUMENT. Is there something else you'd like me to use? https://github.com/project-chip/connectedhomeip/blob/4f0ade5fcd50f99f53188ec401e18c2a22f47e14/src/app/util/af-enums.h#L23

File: src/app/clusters/ota-server/ota-server.cpp
Comment (holbrookt) [reply src/app/clusters/ota-server/ota-server.cpp:170]: I don't know the answer to this, I've just seen Network Commissioning Cluster do the same. I'll see what I can find out about this

File: src/app/clusters/ota-server/ota-server.cpp
Comment (holbrookt) [reply src/app/clusters/ota-server/ota-server.cpp:149]: CHAR_STRING is not yet supported in xml. See #6112

Comment (holbrookt) [reply src/app/clusters/ota-server/ota-server.cpp:149]: Another option would be to use OCTET_STRING for now, which is supported (this is suggested in #6805)

File: src/app/clusters/ota-server/ota-server.cpp
Comment (tcarmelveilleux) [reply src/app/clusters/ota-server/ota-server.cpp:107]: You have to add it, since Ember library predates Matter and doesn't have it yet.

File: src/app/clusters/ota-server/ota-server.cpp
Comment (tcarmelveilleux) [reply src/app/clusters/ota-server/ota-server.cpp:149]: Please add a TODO and open an issue to track making this right once CHAR_STRING is implemented.

File: src/app/clusters/ota-server/ota-server.cpp
Comment (holbrookt) [reply src/app/clusters/ota-server/ota-server.cpp:107]: I can add it.

Just to clarify - are you suggested that EMBER_STATUS_ errors also reflect all CHIP_ERROR_ types?

Comment (holbrookt) [reply src/app/clusters/ota-server/ota-server.cpp:107]: Also, what should EMBER_ZCL_STATUS_INVALID_FIELD be used for then?

File: src/app/clusters/ota-server/ota-server.cpp
Comment (holbrookt) [reply src/app/clusters/ota-server/ota-server.cpp:149]: Filed #7112

File: src/app/clusters/ota-server/ota-server.cpp
Comment (holbrookt) [reply src/app/clusters/ota-server/ota-server.cpp:170]: Should we differentiate by Endpoint as well? Just in case there's an application that wants multiple OTA Server instances?

**Comment by holbrookt** [comment]
Apologies for the messy PR, I've been struggling with a few ZAP bugs and merge conflicts (see https://github.com/project-chip/zap/issues/158 and #7135).

Hand-modified files are:
```
src/app/clusters/ota-server/ota-server-delegate.h
src/app/clusters/ota-server/ota-server.cpp
src/app/clusters/ota-server/ota-server.h
src/app/zap-templates/zcl/chip-ota.xml
src/app/zap_cluster_list.py
src/darwin/Framework/CHIP/templates/helper.js
```

And modified zap files are:
```
examples/all-clusters-app/all-clusters-common/all-clusters-app.zap
src/controller/data_model/controller-clusters.zap
```

**Comment by holbrookt** [comment]
> Per the updated template, can you update the PR here?

@woody-apple Please see updated description.

File: examples/chip-tool/commands/clusters/Commands.h
Comment (tcarmelveilleux) [inline examples/chip-tool/commands/clusters/Commands.h:12064]: `strlen` should not be called on mMetadataForServer

File: examples/chip-tool/commands/clusters/Commands.h
Comment (tcarmelveilleux) [inline examples/chip-tool/commands/clusters/Commands.h:12081]: This is not a char*, it is an octet string, which may likely contain internal nulls

File: src/app/clusters/ota-server/ota-server-delegate.h
Comment (tcarmelveilleux) [inline src/app/clusters/ota-server/ota-server-delegate.h:43]:
```suggestion
    virtual ~OTAServerDelegate() = default;
```

File: src/app/clusters/ota-server/ota-server.cpp
Comment (tcarmelveilleux) [inline src/app/clusters/ota-server/ota-server.cpp:90]: Suggest you make the server log errors with enough details to debug.

**Review by tcarmelveilleux — APPROVED** [review]

File: src/app/clusters/ota-server/ota-server.cpp
Comment (holbrookt) [reply src/app/clusters/ota-server/ota-server.cpp]: Got it, I defined a new error and I'll set it to 0xC6

File: examples/chip-tool/commands/clusters/Commands.h
Comment (holbrookt) [reply examples/chip-tool/commands/clusters/Commands.h:12081]: @mrjerryjohns says this is intended (for now?).

Comment (holbrookt) [reply examples/chip-tool/commands/clusters/Commands.h:12081]: Not sure about internal nulls though

File: src/app/clusters/ota-server/ota-server.cpp
Comment (holbrookt) [reply src/app/clusters/ota-server/ota-server.cpp:90]: Will do

**Review by woody-apple — APPROVED** [review]

**Review by jmartinez-silabs — APPROVED** [review]

**Review by andy31415 — APPROVED** [review]

**Comment by holbrookt** [comment]
I believe this PR is being blocked by #7135

**Comment by woody-apple** [comment]
Seems to work after your last revert. Kicking ESP32 again...

**Comment by woody-apple** [comment]
Force merging, looks like something is wrong with the ESP32 artifact download..

**Comment by woody-apple** [comment]
Reverting, looks like this needs a re-base and fixes :(

**Pull request merged**
